hikari/api/interaction_server.py

```diff
@@ -0,0 +1,186 @@
+# -*- coding: utf-8 -*-
+# cython: language_level=3
+# Copyright (c) 2020 Nekokatt
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""Provides an interface for Interaction REST server API implementations to follow."""
+from __future__ import annotations
+
+__all__: typing.Sequence[str] = ["MainListenerT", "Response", "InteractionServer"]
+
+import abc
+import typing
+
+if typing.TYPE_CHECKING:
+    from hikari.api import special_endpoints
+    from hikari.events import interaction_events
+
+
+MainListenerT = typing.Callable[
+    ["interaction_events.InteractionCreateEvent"],
+    typing.Awaitable[typing.Optional["special_endpoints.InteractionResponseBuilder"]],
+]
+"""Type hint of a Interaction server's listener callback.
+
+This should be an async callback which takes in one positional argument of type
+`hikari.events.interaction_events.InteractionCreateEvent` and may return an
+instance of `hikari.api.special_endpoints.InteractionResponseBuilder` which will
+instruct the server on how to respond or `builtins.None` to let the server
+default to a `hikari.interactions.InteractionResponseType.ACKNOWLEDGE` response.
+
+!!! note
+    For the standard implementation of
+    `hikari.api.special_endpoints.InteractionResponseBuilder` see
+    `hikari.impl.special_endpoints.InteractionResponseBuilder`
+"""
+
+
+class Response(typing.Protocol):
+    """Protocol of the data returned by `InteractionServer.on_interaction`.
```

Comment:
Should we rename this and make it an abc? You seem to be using abc.abstractmethod, so im going to guess its meant to be that